File: daily_problems/python/june_18.py
#!/usr/bin/env python3

# This problem was asked by Google.

# A unival tree (which stands for "universal value") is a tree where all nodes
# under it have the same value.

# Given the root to a binary tree, count the number of unival subtrees.

# For example, the following tree has 5 unival subtrees:

#    0
#   / \
#  1   0
#     / \
#    1   0
#   / \
#  1   1

class Node:
    def __init__(self, val, lc=None, rc=None):
        self.val = val
        self.left = lc
        self.right = rc

def count_subtrees(root):
    if not (root.left or root.right):
        return (root.val, 1)
    
    left_st, left_count = count_subtrees(root.left)
    right_st, right_count = count_subtrees(root.right)
    if left_st == right_st == root.val:
        return (root.val, left_count + right_count + 1)
    else:
        return (root.val, left_count + right_count)

def main():
    test_tree = Node(0, Node(1), Node(0, Node(1, Node(1), Node(1)), Node(0)))
    res = count_subtrees(test_tree)
    print('Result is: %d' % res[1])

if __name__ == '__main__':
    main()

File: daily_problems/haskell/june_18.hs
#!/usr/bin/env runhaskell

-- This problem was asked by Google.
--
-- A unival tree (which stands for "universal value") is a tree where all nodes
-- under it have the same value.
--
-- Given the root to a binary tree, count the number of unival subtrees.
--
-- For example, the following tree has 5 unival subtrees:
--
--    0
--   / \
--  1   0
--     / \
--    1   0
--   / \
--  1   1

data Node
    = Node Int Node Node
    | Leaf Int

countSubtrees :: Node -> (Int, Int)
countSubtrees (Leaf v) = (v, 1)
countSubtrees (Node v lc rc) = (v, newCount)
    where
        (ltVal, leftCount) = countSubtrees lc
        (rtVal, rightCount) = countSubtrees rc 
        newCount = if ltVal == v && rtVal == v 
                    then leftCount + rightCount + 1
                    else leftCount + rightCount

main :: IO ()
main = do
    let tr = Node 0 (Leaf 1) (Node 0 (Node 1 (Leaf 1) (Leaf 1)) (Leaf 0))
        res = countSubtrees tr
    putStrLn $ "Result is "++((show . snd) res)